### PR TITLE
APPS-1794 add instance type by exec

### PR DESCRIPTION
**Pull request by YuxinShi0423**
This PR is to:

- add '--instance-type-by-executable' arg for `dx run` and `dx-jobutil-job-new`
- add corresponding field `system_requirements` and `system_requirements_by_executable` to `DXExecution.run()` and `DXJob.new()`
- fix APPS-1804 that fpga driver setting is missing added to system requirement mapping

### CODE REVIEW

**Comment by YuxinShi0423** [comment]
https://jenkins-vstg.internal.dnanexus.com/view/Traceability%20Platform%20Tests/job/RTM/job/dx-toolkit_traceability_tests_isolated/1506/

File: src/python/dxpy/bindings/dxapplet.py [hunk 1]
```diff
@@ -184,9 +191,11 @@ def _get_cleanup_keys(self):
         raise NotImplementedError('_get_cleanup_keys is not implemented')
 
     def run(self, executable_input, project=None, folder=None, name=None, tags=None, properties=None, details=None,
```
Comment (kpjensen) [inline src/python/dxpy/bindings/dxapplet.py:193]: Any new arguments should be appended to the end of the arguments. This is a breaking change if the caller of the function is not using keyword arguments.

File: src/python/dxpy/bindings/dxjob.py [hunk 2]
```diff
@@ -43,7 +43,7 @@
 #########
 
 def new_dxjob(fn_input, fn_name, name=None, tags=None, properties=None, details=None,
-              instance_type=None, depends_on=None,
+              instance_type=None, cluster_spec=None, fpga_driver=None, system_requirements=None, system_requirements_by_executable=None, depends_on=None,
```
Comment (kpjensen) [inline src/python/dxpy/bindings/dxjob.py:46]: Same argument order comment.

**Review by kpjensen — CHANGES_REQUESTED** [review]

**Comment by YuxinShi0423** [comment]
https://jenkins-vstg.internal.dnanexus.com/view/Traceability%20Platform%20Tests/job/RTM/job/dx-toolkit_traceability_tests_isolated/1509/
https://jenkins-vstg.internal.dnanexus.com/job/dx-toolkit/job/dxpy-branch-integration-tests-on-staging-20.04-python3/496/

File: src/python/dxpy/bindings/dxjob.py (hunk 2 above)
Comment (YuxinShi0423) [reply src/python/dxpy/bindings/dxjob.py]: Thanks for pointing this out :D Fixed.

**Review by wormsik — COMMENTED** [review]
leaving final approval to Kurt, but LGTM 👍

File: src/python/dxpy/scripts/dx.py [hunk 3]
```diff
@@ -3149,24 +3149,61 @@ def run_batch_all_steps(args, executable, dest_proj, dest_path, input_json, run_
...
-                                  **(args.instance_type or {}))
+    requested_instance_type, requested_cluster_spec = {}, {}
+    executable_describe = None
 
-    if args.sys_reqs_from_clone and not isinstance(args.instance_count, str):
-        # extract instance counts from cloned sys reqs and override them with args provided with "dx run"
-        args.instance_count = dict({fn: reqs['clusterSpec']['initialInstanceCount']
-                                        for fn, reqs in list(args.sys_reqs_from_clone.items()) if 'clusterSpec' in reqs},
-                                   **(args.instance_count or {}))
+    from ..utils import merge
```
Comment (wormsik) [inline src/python/dxpy/scripts/dx.py:3155]: I would consider placing this import on the top...or is there any specific reason to have it here? can this import break something (functionality, different OS, ...)?

File: src/python/dxpy/scripts/dx.py [hunk 4]
```diff
@@ -3039,7 +3039,7 @@ def _get_input_for_run(args, executable, preset_inputs=None, input_name_prefix=N
     if args.input_json is None and args.filename is None:
         # --input-json and --input-json-file completely override input
         # from the cloned job
-        exec_inputs.update(args.input_from_clone, strip_prefix=False)
```
Comment (wormsik) [inline src/python/dxpy/scripts/dx.py:3042]: is there any other use of `input_from_clone` field?

File: src/python/dxpy/scripts/dx.py (hunk 3 above)
Comment (YuxinShi0423) [reply src/python/dxpy/scripts/dx.py]: good point! I just feel like it is only used here so better to not import it always. But changed it as you suggested. Thanks!

File: src/python/dxpy/scripts/dx.py [hunk 5]
```diff
@@ -47,14 +47,14 @@
...
-                           instance_type_arg, process_instance_type_arg, process_instance_count_arg, get_update_project_args,
+                           instance_type_arg, process_instance_type_arg, process_instance_type_by_executable_arg, process_instance_count_arg, get_update_project_args,
                            property_args, tag_args, contains_phi, process_phi_param, process_external_upload_restricted_param)
 from ..cli.exec_io import (ExecutableInputs, format_choices_or_suggestions)
 from ..cli.org import (get_org_invite_args, add_membership, remove_membership, update_membership, new_org, update_org,
                        find_orgs, org_find_members, org_find_projects, org_find_apps)
 from ..exceptions import (err_exit, DXError, DXCLIError, DXAPIError, network_exceptions, default_expected_exceptions,
                           format_exception)
-from ..utils import warn, group_array_by_field, normalize_timedelta, normalize_time_input
+from ..utils import warn, group_array_by_field, normalize_timedelta, normalize_time_input, merge
```
Comment (YuxinShi0423) [inline src/python/dxpy/scripts/dx.py:57]: @wormsik import `merge` here

File: src/python/dxpy/scripts/dx.py [hunk 6]
```diff
@@ -3596,10 +3634,9 @@ def run(args):
                 setattr(args, metadata, clone_desc.get(metadata))
 
         if clone_desc['class'] == 'job':
+            args.cloned_job_desc = clone_desc
             if args.executable == "":
                 args.executable = clone_desc.get("applet", clone_desc.get("app", ""))
-            args.input_from_clone = clone_desc["runInput"]
```
Comment (YuxinShi0423) [inline src/python/dxpy/scripts/dx.py:3601]: `args.input_from_clone` is specified from `clone_desc` and used to override input as         ```exec_inputs.update(args.input_from_clone, strip_prefix=False)```

here I made it resolved later from `args.cloned_job_desc` as ```        exec_inputs.update(args.cloned_job_desc.get("runInput", {}), strip_prefix=False)```

@wormsik

**Review by kpjensen — APPROVED** [review]

**Pull request merged**